library/psa_crypto.c

```diff
@@ -4698,21 +4698,107 @@ static psa_status_t psa_generate_derived_key_internal(
     size_t storage_size = bytes;
     psa_status_t status;
 
-    if( ! key_type_is_raw_bytes( slot->attr.type ) )
-        return( PSA_ERROR_INVALID_ARGUMENT );
-    if( bits % 8 != 0 )
-        return( PSA_ERROR_INVALID_ARGUMENT );
-    data = mbedtls_calloc( 1, bytes );
-    if( data == NULL )
-        return( PSA_ERROR_INSUFFICIENT_MEMORY );
+    /*
+    * ECC key types require the generation of a private key which is an integer
+    * in the range [1, N - 1], where N is the boundary of the private key domain:
+    * N is the prime p for Diffie-Hellman, or the order of the
+    * curve’s base point for ECC.
+    *
+    * Let m be the bit size of N, such that 2^m > N >= 2^(m-1).
+    * This function generates the private key using the following process:
+    *
+    * 1. Draw a byte string of length ceiling(m/8) bytes.
+    * 2. If m is not a multiple of 8, set the most significant
+    *    (8 * ceiling(m/8) - m) bits of the first byte in the string to zero.
+    * 3. Convert the string to integer k by decoding it as a big-endian byte string.
+    * 4. If k > N - 2, discard the result and return to step 1.
+    * 5. Output k + 1 as the private key.
+    *
+    * This method allows compliance to NIST standards
+    */
+    if ( PSA_KEY_TYPE_IS_ECC( slot->attr.type ) )
+    {
+        int cmp_result;
+        do {
+            int ret;
+            psa_ecc_family_t curve = PSA_KEY_TYPE_ECC_GET_FAMILY(
+                                        slot->attr.type );
+            mbedtls_ecp_group_id grp_id =
+                mbedtls_ecc_group_of_psa( curve, bits, 0 );
 
-    status = psa_key_derivation_output_bytes( operation, data, bytes );
-    if( status != PSA_SUCCESS )
-        goto exit;
+            mbedtls_ecp_keypair ecp;
+            mbedtls_ecp_keypair_init( &ecp );
+
+            if( ( ret = mbedtls_ecp_group_load( &ecp.grp, grp_id ) ) != 0 )
```

Comment:
Aw, it's pretty awkward that we have do to this in order get the value of `N` for this curve, but I don't see a better way. We should be doing this outside of the loop, though.